include/deal.II/grid/tria_accessor.h

```diff
@@ -5698,9 +5645,29 @@ TriaAccessor<structdim, dim, spacedim>::set_line_orientation(
 {
   Assert(used(), TriaAccessorExceptions::ExcCellNotUsed());
   AssertIndexRange(line, this->n_lines());
+  Assert(dim != 1,
+         ExcMessage("In 1d lines are cells and thus do not need to have their "
+                    "orientations set."));
+  Assert(dim != 2,
+         ExcMessage("In 2d lines are faces, and, for compatibility with other "
+                    "dimensions, their orientations should be set via "
+                    "set_combined_orientation()."));
+  // work around a bogus GCC-9 warning which considers line and value unused
+  // except in 3d
+  (void)line;
+  (void)value;
 
-  dealii::internal::TriaAccessorImplementation::Implementation::
-    set_line_orientation(*this, line, value);
+  if constexpr (dim == 3)
+    {
+      // We set line orientations per face, not per cell, so this only works for
+      // faces in 3d
+      Assert(structdim == 2, ExcNotImplemented());
+      Assert(this->present_index * GeometryInfo<3>::lines_per_face + line <
+               this->tria->faces->quads_line_orientations.size(),
+             ExcInternalError());
+      this->tria->faces->quads_line_orientations
+        [this->present_index * GeometryInfo<3>::lines_per_face + line] = value;
```

Comment:
I assume this function is only called for hex meshes? Perhaps that would be worth assertion.

Reply:
This is one of those cases where by `GeometryInfo` we really mean `MaximumGeometryInfo`: i.e., that array always has size `n_cells() * GeometryInfo<3>::faces_per_cell * Geometryinfo<3>::lines_per_face` and we just ignore half of it if we have a purely tetrahedral mesh.

I'll add a todo item to the orientation issue since we do this a lot in the orientation code and it requires knowing that context.

Reply:
OK, thanks for the explanation. Let's merge. If you wanted to add a comment in these places (can be the same comment in all), then I would accept such a patch too.

Reply:
I should add that ultimately, we want to get rid of `GeometryInfo` altogether. If, for future patches, you want to use `ReferenceCells::hyper_cube<dim>()`, then that would avoid having to touch these places again.